crates/events-dispatcher/src/dispatcher.rs

```diff
@@ -0,0 +1,246 @@
+    }
+}
+
+impl<T: Eq> PartialOrd for Scheduled<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct EventsDispatcherCfg {
+    pub sources: Vec<Inlet<PeerEvent>>,
+}
+
+pub struct EventsDispatcher {
+    sources: Vec<Inlet<PeerEvent>>,
+    recv_cmd_channel: Inlet<Command>,
+    send_events: Outlet<ListenerEvent>,
+}
+
+impl EventsDispatcher {
+    pub fn new(cfg: EventsDispatcherCfg) -> (Self, EventsDispatcherApi, Inlet<ListenerEvent>) {
+        let (send_cmd_channel, recv_cmd_channel) = unbounded();
+        let api = EventsDispatcherApi { send_cmd_channel };
+
+        let (send_events, recv_events) = unbounded();
+
+        let this = Self {
+            sources: cfg.sources,
+            recv_cmd_channel,
+            send_events,
+        };
+        (this, api, recv_events)
+    }
+
+    pub fn start(self) -> task::JoinHandle<()> {
+        task::spawn(self.run())
+    }
+
+    async fn run(self) {
+        let send_events = self.send_events;
+
+        let mut recv_cmd_channel = self.recv_cmd_channel.fuse();
+        let sources = self
+            .sources
+            .into_iter()
+            .map(|source| source.fuse())
+            .collect::<Vec<_>>();
+        let mut sources_channel = stream::select_all(sources);
+
+        let mut heap: BinaryHeap<Scheduled<String>> = BinaryHeap::new();
+        let mut listeners = Listeners::new();
+
+        loop {
+            let now = Instant::now();
+            let mut timer = {
+                let next_task_in = if let Some(next_task) = heap.peek() {
+                    next_task
+                        .run_at
+                        .checked_duration_since(now)
+                        .unwrap_or_default()
+                } else {
+                    Duration::MAX
+                };
+
+                async_std::stream::interval(next_task_in).fuse()
+            };
+
+            select! {
+                command = recv_cmd_channel.select_next_some() => {
+                    match command {
+                        Command::Add(listener, config) => {
+                            let listener = Arc::new(listener);
+                            for config in config.configs {
+                                match config {
+                                    EventConfig::Timer(config) => {
+                                        let periodic = Periodic { id: listener.id.clone(), period: config.period };
+                                        heap.push(Scheduled::new(periodic, Instant::now()))
+                                    },
+                                    EventConfig::PeerEvent(config) => {
+                                        listeners.add(listener.clone(), config.events);
+                                    },
+                                }
+                            }
+                        },
+                        Command::Remove(listener_id, reply) => {
```

Comment:
OneshotOutlet is usually called `out` everywhere (eg ConnectionPoolApi and KademliaApi), let's stick to consistent naming 

```suggestion
                        Command::Remove(listener_id, out) => {
```